### DISCUSSION TITLE
How can I call an async function in iOS custom plugin?

### DISCUSSION THREAD

**Discussion by DaDlugosch** [ionic-team/capacitor#7282] [Q&A]
Hi, I've created a custom iOS plugin which I inject and call inside my JS function. I would like to use it as an oauth authentication middleware, however I need to call `try await` inside it and based on that the function must be `async`. But when I set an async keyword to my function: `@objc func echo(_ call: CAPPluginCall) async {}` after calling it inside my app I've got this error response: 
```bash 
Error: Plugin MyCustomPlugin does not respond to method call "echo" using selector "echo:".
Ensure plugin method exists, uses @objc in its declaration, and arguments match selector without callbacks in CAP_PLUGIN_METHOD.
```
When I call this method without `async` keyword, everything works (except `try await` function call, of course)

**Comment by 0x7061** [comment]
Good point. This is also a limitation on the Android side. Unfortunately you can't have the `@PluginMethod` decorator on a suspending function.

One way to solve this would be to switch into an async context within the synchronous function. So in iOS you could invoke a Task where you process the work and resolve the plugin call.
```swift
@objc
func mySynchronousPluginMethod(_ call: CAPPluginCall) {
    Task {
        let success = await myAsyncMethod()
        if (success) {
           call.resolve()
        } else {
           call.reject("my error")
        }
    }
}
```
Android equivalent might look like this.
```kotlin
@PluginMethod
fun mySynchronousPluginMethod(call: PluginCall) {
    GlobalScope.launch {
        val success = mySuspendingMethod()
        if (success) {
            call.resolve()
        } else {
            call.reject("my error")
        }
    }
}
```
